### PR TITLE
Added google call creds/per_rpc interop tests

### PR DESCRIPTION
**Pull request by kpayson64**
This is a breakout of the per_rpc_creds from #6155.
Adds a wrapper for google creds, as well as implementing the per_rpc interop tests in issue #4389

### CODE REVIEW

File: src/python/grpcio/grpc/beta/_auth.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from concurrent.futures import ThreadPoolExecutor
+
+from grpc.beta import interfaces
+
+class GoogleCallCredentials(interfaces.GRPCAuthMetadataPlugin):
+  """ Metadata wrapper for googleCredentials from the oauth2client library
```
Comment (jtattermusch) [inline src/python/grpcio/grpc/beta/_auth.py]: nit: I think the comment should be """Some comment with no spaces around""", if it fits on one line. Also below.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 2]
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
     A CallCredentials object for use in a GRPCCallOptions object.
   """
   if name is None:
-    name = metadata_plugin.__name__
+    name = metadata_plugin.__class__.__name__
   return CallCredentials(
       _low.call_credentials_metadata_plugin(metadata_plugin, name))
 
+def google_call_credentials(credentials):
```
Comment (jtattermusch) [inline src/python/grpcio/grpc/beta/implementations.py]: I think the naming "google_call_credentials" and "access_token_call_credential" is good, let's get a second opinion from @nathanielmanistaatgoogle  or @soltanmm.  If everything else looks good, I think it's fine to submit this change and revisit the naming if necessary.

**Comment by jtattermusch** [comment]
LGTM after fixing the comment format I mentioned.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# Copyright 2015, Google Inc.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: Header copied from last year.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from concurrent.futures import ThreadPoolExecutor
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: ["Use imports for packages and modules only."](https://google.github.io/styleguide/pyguide.html?showone=Imports#Imports) Do not import classes, functions, or other code elements of modules.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  def __init__(self, access_token):
+    self._access_token = access_token
+
+  def __call__(self, context, callback):
+    _sign_request(callback, self._access_token, None)
+
+
+def _sign_request(callback, token, error):
+  metadata = [('authorization', 'Bearer {}'.format(token))]
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: Never write a list literal (`[<expression>]`) where a tuple literal will do (`(<expression>,)`). Will a tuple literal do here?

File: src/python/grpcio/grpc/beta/_auth.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  """Metadata wrapper for raw access token credentials"""
+
+  def __init__(self, access_token):
+    self._access_token = access_token
+
+  def __call__(self, context, callback):
+    _sign_request(callback, self._access_token, None)
+
+
+def _sign_request(callback, token, error):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: Please place this function earlier in the file than ("above") all of its use-sites.

File: src/python/grpcio/tests/interop/methods.py [hunk 7]
```diff
@@ -39,12 +39,14 @@
 
 from oauth2client import client as oauth2client_client
 
+from grpc.beta import interfaces, implementations
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/interop/methods.py]: One import per line, please.

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
+# Copyright 2015, Google Inc.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: If you authored this file this year, please date its header this year. If this file is a branch of an earlier file, should the code review tool be showing me something on the left side of this diff?

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+class AccessTokenCallCredentialsTest(unittest.TestCase):
+  
+  def test_google_call_credentials_success(self):
+    callback_event = threading.Event()
+    test = self
+
+    class MockPluginCallback:
+      def __call__(self, metadata, error):
+        test.assertEqual(metadata, [('authorization', 'Bearer token')])
+        test.assertEqual(error, None)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: `assertIsNone` (or `assertIsNotNone`) when making comparisons with `None`.

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+      def __call__(self, metadata, error):
+        test.assertIsNotNone(error)
+        callback_event.set()
+
+    class MockGoogleCreds:
+      def get_access_token(self):
+        raise Exception()
+
+    call_creds = _auth.GoogleCallCredentials(MockGoogleCreds())
+    call_creds.__call__(None, MockPluginCallback())
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: `call_creds(None, MockPluginCallback())`?

**Comment by kpayson64** [comment]
I've addressed all the comments.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""GRPCAuthMetadataPlugins for standard authentication."""
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: Blank line between license header and module doc string.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+import concurrent.futures
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: `from concurrent import futures`.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+from grpc.beta import interfaces
+
+
+def _sign_request(callback, token, error):
+  metadata = (('authorization', 'Bearer {}'.format(token)),)
+  callback(metadata, error)
+
+
+class GoogleCallCredentials(interfaces.GRPCAuthMetadataPlugin):
+  """Metadata wrapper for googleCredentials from the oauth2client library."""
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: Is the code element to which you're referring actually styled `googleCredentials`? Maintain consistent styling (case, underscores, and so forth) when referring to a code element in documentation.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 14]
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
...
+  """Construct CallCredentials from google credentials
+
+  Args:
+    credentials: A google credentials object from oauth2client
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: Don't be ambiguous about the type of this parameter - use the correct, formal name of the interface or class.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 15]
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
     A CallCredentials object for use in a GRPCCallOptions object.
   """
   if name is None:
-    name = metadata_plugin.__name__
+    name = metadata_plugin.__class__.__name__
   return CallCredentials(
       _low.call_credentials_metadata_plugin(metadata_plugin, name))
 
+def google_call_credentials(credentials):
+  """Construct CallCredentials from google credentials
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: Missing period at the end.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 16]
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
...
+
+  Args:
+    credentials: A google credentials object from oauth2client
+
+  Returns:
+    A CallCredentials object for header authentication
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: Missing period at the end.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 17]
```diff
@@ -105,10 +105,33 @@ def metadata_call_credentials(metadata_plugin, name=None):
...
+  """
+  return metadata_call_credentials(_auth.GoogleCallCredentials(credentials))
+
+def access_token_call_credentials(access_token):
+  """Construct CallCredentials from a raw access token
+
+  Args:
+    access_token: A raw access token used for authentication
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: What's a "raw access token"? A byte string? An instance of some interface or class?

File: src/python/grpcio/tests/unit/beta/_implementations_test.py [hunk 18]
```diff
@@ -48,6 +48,19 @@ def test_application_provided_root_certificates(self):
     self.assertIsInstance(
         channel_credentials, implementations.ChannelCredentials)
 
+class CallCredentialsTest(unittest.TestCase):
+
+  def test_google_call_credentials(self):
+    class MockGoogleCreds:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_implementations_test.py]: Missing inheritance from `object` - but why are you using a mock here anyway? Why not use a real instance?

[I have some opinions on the subject](https://www.youtube.com/watch?v=Xu5EhKVZdV8).

File: src/python/grpcio/tests/tests.json [hunk 19]
```diff
@@ -50,4 +53,4 @@
   "cygrpc_test.InsecureServerInsecureClient", 
   "cygrpc_test.SecureServerSecureClient", 
   "cygrpc_test.TypeSmokeTest"
-]
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/tests.json]: I suspect that you didn't intend the change on this line?

File: src/python/grpcio/tests/interop/client.py [hunk 20]
```diff
@@ -65,39 +65,35 @@ def _args():
...
+          channel_creds, call_creds)
+
     channel = test_utilities.not_really_secure_channel(
-        args.server_host, args.server_port,
-        implementations.ssl_channel_credentials(root_certificates),
+        args.server_host, args.server_port, channel_creds,
         args.server_host_override)
     stub = test_pb2.beta_create_TestService_stub(
-        channel, metadata_transformer=metadata_transformer)
+        channel)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/interop/client.py]: This line-wrap is now unnecessary, right?

**Comment by kpayson64** [comment]
I've addressed the comments.

File: src/python/grpcio/tests/interop/client.py [hunk 21]
```diff
@@ -109,8 +104,7 @@ def _test_case_from_arg(test_case_arg):
   for test_case in methods.TestCase:
     if test_case_arg == test_case.value:
       return test_case
-  else:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/interop/client.py]: Why this change? I prefer the `for`-`else`.

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+      callback_event.set()
+
+    class MockAccessToken(dict):
+      pass
+
+    class MockGoogleCreds(object):
+
+      def get_access_token(self):
+        token = MockAccessToken()
+        token.access_token = 'token'
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: Concept/abstraction weirdness: why does something that is an access token have an `access_token` attribute? Access tokens aren't structurally recursive, are they?

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+    self.assertTrue(callback_event.wait(1.0))
+
+  def test_google_call_credentials_error(self):
+    callback_event = threading.Event()
+
+    def mock_callback(metadata, error):
+      self.assertIsNotNone(error)
+      callback_event.set()
+
+    class MockGoogleCreds(object):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: This class doesn't require any of what's available to it in function-scope; please extract it to module-scope.

File: src/python/grpcio/grpc/beta/implementations.py [hunk 24]
```diff
@@ -105,10 +106,37 @@ def metadata_call_credentials(metadata_plugin, name=None):
     A CallCredentials object for use in a GRPCCallOptions object.
   """
   if name is None:
-    name = metadata_plugin.__name__
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: I don't think I understand the change on this line? What's going on here?

Comment (kpayson64) [inline src/python/grpcio/grpc/beta/implementations.py]: By default python objects don't have a __name__ attribute.  If I pass in an arbitrary MetadataPlugin, it will throw an exception.

File: src/python/grpcio/tests/interop/client.py (hunk 21 above)
Comment (kpayson64) [inline src/python/grpcio/tests/interop/client.py]: glint complained:
W0120:107:_test_case_from_arg: Else clause on loop without a break statement [useless-else-on-loop]

I've changed it back though.

**Comment by kpayson64** [comment]
I've addressed the feedback.

File: src/python/grpcio/grpc/beta/implementations.py (hunk 24 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/implementations.py]: It seems to me that an object's name, if present, is more likely to be helpful than its class name. Should this code look for the object name and only use the class name if it is not found? I believe @soltanmm originally authored this and may have an opinion.

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+class MockGoogleCreds(object):
+
+  def get_access_token(self):
+    token = MockAccessTokenInfo()
+    token.access_token = 'token'
+    return token
+
+
+class MockExcepitonGoogleCreds(object):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: Typo "Excepiton".

File: src/python/grpcio/grpc/beta/implementations.py (hunk 24 above)
Comment (kpayson64) [inline src/python/grpcio/grpc/beta/implementations.py]: @soltanmm 
As far as I can tell, objects having names is the exception and not the rule.  Is that the behavior we want, or are we just obfuscating naming unnecessarily?

Comment (soltanmm) [inline src/python/grpcio/grpc/beta/implementations.py]: Woops, this one slipped by me. I think your change is the right way to go for non-function callables (which I think I'd not considered at the time). The problem is that we expect most users to just write a new function and pass it in. Functions have `__name__` attributes that are meaningful, but their `__class__` attributes are always `function`. So we'd at least want to maintain support for that use case.

Comment (soltanmm) [inline src/python/grpcio/grpc/beta/implementations.py]: And, yes, that means we're supporting something that doesn't directly inherit from an interface we're requesting, and instead are duck typing, and haven't taken the time to do some ABC tricks to make it airtight, but it's straightforward enough that in offline discussion we determined it was desirable to implicitly support it.

Comment (kpayson64) [inline src/python/grpcio/grpc/beta/implementations.py]: I've updated the behavior to check for the function name first, and then fall back on the class name.

**Comment by kpayson64** [comment]
test this please

**Comment by kpayson64** [comment]
test this please

**Comment by kpayson64** [comment]
test this please

**Comment by kpayson64** [comment]
I've addressed the feedback, and also fixed a bug where MetadataPlugin destruction was causing a intermittent segfaults, which caused the interop tests to fail.

File: src/python/grpcio/grpc/beta/_auth.py [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+    self._credentials = credentials
+    self._pool = futures.ThreadPoolExecutor(max_workers=1)
+
+  def __call__(self, context, callback):
+    # MetadataPlugins cannot block (see grpc.beta.interfaces.py)
+    future = self._pool.submit(self._credentials.get_access_token)
+    future.add_done_callback(lambda x: self._get_token_callback(callback, x))
+
+  def _get_token_callback(self, callback, future):
+    access_token = None
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/beta/_auth.py]: No need for these fields at method top-level - use control flow instead:
```
try:
  access_token = future.result().access_token
except Exception as e:
  _sign_request(callback, None, e)
else:
  _sign_request(callback, access_token, None)
```
.

File: src/python/grpcio/tests/unit/beta/_auth_test.py [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+
+"""Tests of standard AuthMetadataPlugins."""
+
+import threading
+import unittest
+
+from grpc.beta import _auth
+
+
+class MockAccessTokenInfo(dict):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/beta/_auth_test.py]: Why does this class need to exist? Why can't a `dict` be used instead at its one use-site?

**Comment by nathanielmanistaatgoogle** [comment]
Looks good to me apart from two comments; @soltanmm please also review.

**Comment by kpayson64** [comment]
Made suggested changes and rebased.

**Comment by nathanielmanistaatgoogle** [comment]
Luhfullyguhtum.

**Comment by kpayson64** [comment]
@soltanmm Friendly ping

File: src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi [hunk 28]
```diff
@@ -137,7 +137,7 @@ cdef void plugin_get_metadata(
   cy_context.context = context
   self.plugin_callback(cy_context, python_callback)
 
-cdef void plugin_destroy_c_plugin_state(void *state):
+cdef void plugin_destroy_c_plugin_state(void *state) with gil:
```
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi]: Sorry for the long round trip. What's up with the added GIL acquisition? Does the decref go boom without it? I could've sworn Cython was supposed to whine if acquiring the GIL was needed to avoid going boom.

Comment (kpayson64) [inline src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi]: I'm not sure about Cython's behavior, but it seems like manipulating python objects in c extensions requires the gil

https://docs.python.org/2/c-api/init.html#thread-state-and-the-global-interpreter-lock

The error I was seeing was a consistent segfault on the first execution of the interop client in a fresh docker instance, and this change fixed it.

Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi]: Huh. Arguably Cython's supposed to statically error out if such a thing is necessary and not present, so that's something to file on their tracker at some point, methinks...

But otherwise, yeah, works for me.

Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi]: Actually, this may not be one for the Cython tracker. May need to investigate a little.

Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi]: Ah. The problem is that Cython correctly sees that, from Python code, this function is only ever called when the GIL is held (due to its old signature implicitly requesting that calling code hold the GIL, else it'd be `nogil`), but we let this function's pointer escape into a non-Python non-Cython function, so the transpiler can't statically enforce that guarantee.

**Comment by soltanmm** [comment]
lgtm on tests

**Comment by kpayson64** [comment]
test this please

**Comment by kpayson64** [comment]
All relevant tests have passed.
